content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+    "startLoc": {
+      "pc": 4386096,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 110,
+      "function": {
+        "name": "runtime.main",
+        "value": 4386096,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  {
+    "id": 2,
+    "currentLoc": {
+      "pc": 4387627,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 303,
+      "function": {
+        "name": "runtime.gopark",
+        "value": 4387392,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "userCurrentLoc": {
+      "pc": 4387627,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 303,
+      "function": {
+        "name": "runtime.gopark",
+        "value": 4387392,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "goStatementLoc": {
+      "pc": 4387029,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 240,
+      "function": {
+        "name": "runtime.init.4",
+        "value": 4386976,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "startLoc": {
+      "pc": 4387056,
+      "file": "/usr/local/go/src/runtime/proc.go",
+      "line": 243,
+      "function": {
+        "name": "runtime.forcegchelper",
+        "value": 4387056,
+        "type": 0,
+        "goType": 0,
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  ...
+```
+
+Every object in the JSON represents a single goroutine. By checking [“goroutines” command help](https://github.com/derekparker/delve/blob/master/Documentation/cli/README.md#goroutines) we can figure out what data Delve knows. We're interested in `userCurrentLoc` field, which is "topmost stackframe in user code".
+
+Of all goroutines in the JSON let's list unique function names with the exact line numbers where function got paused:
+
+```
+= jq -c '.result[] | [.userCurrentLoc.function.name, .userCurrentLoc.line]' server-test-1_dlv-rpc-list_goroutines.json | sort | uniq -c
+
+   1 ["internal/poll.runtime_pollWait",173]
+1000 ["main.(*Metrics).CountS",113]
+   1 ["main.(*Metrics).SetM",145]
+   1 ["main.(*Metrics).startOutChannelConsumer",239]
+   1 ["main.run",56]
+   1 ["os/signal.signal_recv",139]
+   6 ["runtime.gopark",303]
+```
+
+The majority of goroutines (1000) have stacked in `main.(*Metrics).CountS:113`. Now, this is the perfect time to look at the source code.
+
+In the `main` package, find `Metrics` struct and look at its `CountS` method (see `example/server/metrics.go`):
+
+```go
+// CountS increments counter per second.
+func (m *Metrics) CountS(key string) {
+    // ···
+
+    m.inChannel <- NewCountMetric(key, 1, second)
+}
+```
+
+Our server has stacked on sending to the `inChannel` channel. Let’s find out who is supposed to read from this channel. After inspecting the code, we should find the following function (example/server/metrics.go):
```

Comment:
* Format the path the same way as the one above
 * Also add a link here? https://github.com/narqo/postmortem-debug-go/blob/2c42ca73ebd500fe8da1c6ac8ecaf4af143aca78/example/server/metrics.go#L110

```suggestion
Our server has stacked on sending to the `inChannel` channel. Let’s find out who is supposed to read from this channel. After inspecting the code, we should find the following function (in `example/server/metrics.go`):
```